src/components/workflow/workflow.service.ts

```diff
@@ -0,0 +1,866 @@
+            relation('out', '', 'workflow', {
+              active: true
+            }),
+            node('workflow'),
+            relation('out', '', 'possibleState', {
+              active: true,
+            }),
+            node('state', 'State', {
+              id: stateId
+            })
+          ]
+        ])
+        .detachDelete('state')
+        .first();
+
+    } catch (e) {
+      this.logger.warning('Failed to delete state', {
+        exception: e,
+      });
+
+      throw e;
+    }
+  };
+
+  // we don't need to have a list workflow function when we have a list state function that takes the baseNodeId // listAllStatesOnWorkflow
+  async listStates(session: ISession, baseNodeId: string): Promise<StateListOutput>{
+    try {
+      const result = (await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true
+            }),
+            node('user'),
+            relation('in', '', 'admin', {
+              active: true
+            }),
+            node('baseNode', 'BaseNode', {
+              id: baseNodeId
+            }),
+            relation('out', '', 'workflow', {
+              active: true
+            }),
+            node('workflow'),
+            relation('out', '', 'possibleState', {
+              active: true,
+            }),
+            node('state')
+          ]
+        ])
+        .return({
+          state:[
+            { id: 'id' },
+            { value: 'value' }
+          ]
+        })
+        .run()) as State[];
+
+      return { items: result.filter(item => item.id && item.value) };
+
+    } catch (e) {
+      this.logger.warning('Failed to delete state', {
+        exception: e,
+      });
+
+      throw e;
+    }
+  };
+
+  // this will be used to get the next possible states of any state, including the current state  // listNextPossibleStates
+  async listNextStates(_session: ISession, stateId: string): Promise<StateListOutput>{
+    try {
+      // WIP, not sure how to check session in this function
+      const result = (await this.db
+        .query()
+        .match([
+          [
+            node('state', 'State', {
+              id: stateId
+            }),
+            relation('out', '', 'nextPossibleState', {
+              active: true
+            }),
+            node('nextState')
+          ]
+        ])
+        .return({
+          nextState:[
+            { id: 'id' },
+            { value: 'value' }
+          ]
+        })
+        .run()) as State[];
+
+      return { items: result.filter(item => item.id && item.value) };
+
+    } catch (e) {
+      this.logger.warning('Failed to delete state', {
+        exception: e,
+      });
+
+      throw e;
+    }
+  };
+
+  // attachSecurityGroupToState
+  async attachSecurityGroup(session: ISession, input: GroupState): Promise<void>{
```

Comment:
the requesting user needs to be an admin of the base node and an admin of the security group in order to attached a security group to a state.